Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java

```diff
@@ -0,0 +1,124 @@
+
+  public ServerHandler(Integer serverSocketPort) {
+    this.serverSocketPort = serverSocketPort;
+  }
+
+  public long getHostSyncTimestamp() {
+    return hostSyncTimestamp;
+  }
+
+  public Integer getServerSocketPort() {
+    return serverSocketPort;
+  }
+
+  public void setServerSocketPort(Integer serverSocketPort) {
+    this.serverSocketPort = serverSocketPort;
+  }
+
+  public long getServerStartTimestamp() {
+    return serverStartTimestamp;
+  }
+
+  public void startConnection() {
+    ConnectionThread connectionThread = new ConnectionThread();
+    connectionThread.execute();
+  }
+
+  public void sendKeySimulationSignal() {
+    hostSyncTimestamp = System.currentTimeMillis();
+    outputWriter.write("started capture*");
+    outputWriter.flush();
+  }
+
+  @NotNull
+  public long getSyncOffset() {
+    return serverStartTimestamp - hostSyncTimestamp;
+  }
+
+  /** Creates a socket for communication with the server. Sets reader and writer for server
+   * connection. */
+  private class ConnectionThread extends AsyncTask<Void, Void, Void> {
+
+    @Override
+    protected Void doInBackground(Void... voids) {
+      try {
+        String SERVER_IP = "127.0.0.1";
+        socket = new Socket(SERVER_IP, serverSocketPort);
+        outputWriter = new PrintWriter(socket.getOutputStream());
+        inputReader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+        Log.d(ContentValues.TAG, "Connection established");
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+      return null;
+    }
+  }
+
+  public ArrayList<Long> downloadServerTimeStamps() {
+    if (serverSocketPort == null || !socket.isConnected()) {
+      Log.d(ContentValues.TAG, "Can't find server to download timestamps. No port bound or connection is dead.");
+      return null;
+    }
+    DownloadServerLogs downloadServerLogsTask = new DownloadServerLogs();
+    try {
+      return downloadServerLogsTask.execute().get();
+    } catch (ExecutionException | InterruptedException e) {
+      e.printStackTrace();
+    }
+    return null;
+  }
+
+  /** Requests the server to send the timestamps of key-presses. */
+  private class DownloadServerLogs extends AsyncTask<Void, Void, ArrayList<Long>> {
+    @Override
+    protected ArrayList<Long> doInBackground(Void... values) {
+      if (serverSocketPort == null) {
+        Log.d(
+            ContentValues.TAG,
+            "No server to read input from. Check if server communication is working.");
+      }
+      ArrayList<Long> serverTimestamps = new ArrayList<>();
+      try {
+        Log.d(ContentValues.TAG, "Sending request to download timestamps");
+        outputWriter.write("send timestamps" + "*");
+        outputWriter.flush();
+        Log.d(ContentValues.TAG, "Request sent to server. Waiting to receive timestamps..");
+        String message = inputReader.readLine();
+        serverStartTimestamp = Long.parseLong(message);
+
+        while (message != null) {
```

Comment:
does this mean the connection is closed after the last timestamp?  How do you send the results back to the server?

Reply:
@mwarton No, the connection is not closed until explicitly done so, or until the `serverHandler` object goes out of scope.
Here it only means that until the inputReader is able to get messages, store that message and read one line at a time from the socket.

Reply:
I guess I'm not clear on why it would return null instead of blocking for more data if the connection is still open.  I assumed it was blocking until the connection was closed by the other side.

Seems odd, but ok for now.

Reply:
Note that the timestamps data is sent from server in a continous stream and this inputReader tries to read it line by line i.e delimiting on `\n` and at the end of that continous stream if would read a null giving us cue that the timestamps are received completely.
Also, I tested with the setup, worked as expected.